### PR TITLE
server & core: handle binary terminal data

### PR DESCRIPTION
**Pull request by bjia56**
Follow up to #1148 

Binary data now treated properly as such. Also optimizes the common data transfer cases by reducing the bytes sent in the stringified message and removing stringification entirely for server-sent data.

### CODE REVIEW

File: plugins/core/ui/src/components/builtin/ShellComponent.vue [hunk 1]
```diff
@@ -40,19 +40,20 @@ export default {
     this.socket.send(JSON.stringify({ dim: { cols: term.cols, rows: term.rows } }));
 
     this.socket.on("message", (data) => {
-      term.write(new Uint8Array(Buffer.from(JSON.parse(data).data)));
+      term.write(new Uint8Array(Buffer.from(data)));
     });
 
     term.onData((data) => {
-      this.socket.send(JSON.stringify({ data }));
+      this.socket.send(JSON.stringify({ d: data }));
```
Comment (koush) [inline plugins/core/ui/src/components/builtin/ShellComponent.vue:47]: Should send this as a buffer as well, all term data can go across the wire as binary. No reason to send json for text.

File: server/src/runtime.ts [hunk 2]
```diff
@@ -119,11 +119,16 @@ export class ScryptedRuntime extends PluginHttp<HttpPluginData> {
                 const spawn = require('node-pty-prebuilt-multiarch').spawn as typeof ptySpawn;
                 const cp = spawn(process.env.SHELL, [], {
                 });
-                cp.onData(data => connection.send(JSON.stringify({data})));
+                cp.onData(data => connection.send(data));
                 connection.on('message', message => {
+                    if (Buffer.isBuffer(message)) {
+                        cp.write(message.toString());
```
Comment (koush) [inline server/src/runtime.ts:125]: Why toString here, can write the buffer?

File: server/src/runtime.ts [hunk 3]
```diff
@@ -119,11 +119,16 @@ export class ScryptedRuntime extends PluginHttp<HttpPluginData> {
...
+                        return;
+                    }
+
                     const parsed = JSON.parse(message.toString());
-                    if (parsed.data) {
-                        cp.write(parsed.data);
+                    if (parsed.d) {
```
Comment (koush) [inline server/src/runtime.ts:130]: This code path will be gone with aforementioned changes

File: server/src/runtime.ts (hunk 2 above)
Comment (bjia56) [reply server/src/runtime.ts:125]: According to this, they are expecting a string https://github.com/microsoft/node-pty/blob/main/typings/node-pty.d.ts#L168

Comment (koush) [reply server/src/runtime.ts:125]: Huh, weird. Looks good then

**Comment by bjia56** [comment]
With the latest changes, core is functionally backwards compatible with existing servers, except it will type in the dimensions into the pty. Server is not backwards compatible with existing core (the console would just stop working, but at least won't crash the server), so it's probably better to release core first, then server shortly afterwards.

File: server/src/runtime.ts [hunk 4]
```diff
@@ -127,9 +127,7 @@ export class ScryptedRuntime extends PluginHttp<HttpPluginData> {
                     }
 
                     const parsed = JSON.parse(message.toString());
```
[the remark this answers is not in the document]
Comment (koush) [reply server/src/runtime.ts]: Oh weird, I must have been looking at an old revision or something, cause it's definitely there now!

**Pull request merged**